craft_application/services/lifecycle.py

```diff
@@ -149,6 +149,7 @@ def __init__(
     def setup(self) -> None:
         """Initialize the LifecycleManager with previously-set arguments."""
         self._lcm = self._init_lifecycle_manager()
+        callbacks.register_prologue(util.set_global_environment)
```

Comment:
Here's a problem:  Callbacks can only be registered once, so applications would have to override `util.set_global_environment()` to add their variables, which isn't ergonomic.

Reply:
ok so the problem is that we need to call a function that configures a ProjectInfo's environment in two places: once when processing the yaml (so even before we have a project) and once when the lifecycle starts up. At the moment I can only think of two solutions, neither of which I'm particularly crazy about:

1) Keep the environment setting in `Application._set_global_environment` and add a new parameter in LifecycleService's constructor like `environment_callback`, that you forward to this call of `register_prologue()`. The application can then set `environment_callback=self._set_global_environment` in `configure_services()` and this lets subclasses override the method;
2) Move all environment handling to a classmethod in the LifecycleService. The application could call something like `self.services.LifecycleClass.set_global_environment()` in its own `_set_global_environment()`, and the tools would be able to override this in their own `LifecycleClass` impl.